lib/livebook/runtime/evaluator.ex

```diff
@@ -620,6 +621,104 @@ defmodule Livebook.Runtime.Evaluator do
     end
   end
 
+  defp eval("erlang", code, binding, env) do
+    try do
+      erl_binding =
+        binding
+        |> Enum.reduce(%{}, fn {name, value}, erl_binding ->
+          :erl_eval.add_binding(elixir_to_erlang_var(name), value, erl_binding)
+        end)
+
+      with {:ok, tokens, _} <- :erl_scan.string(code |> String.to_charlist()),
+           {:ok, parsed} <- :erl_parse.parse_exprs(tokens),
+           {:value, result, new_erl_binding} <- :erl_eval.exprs(parsed, erl_binding) do
+        new_erl_binding =
+          new_erl_binding
+          |> Map.drop(Map.keys(erl_binding))
+          |> Enum.reduce(%{}, fn {name, value}, acc ->
+            Map.put(acc, erlang_to_elixir_var(name), value)
+          end)
+
+        binding =
+          new_erl_binding
+          |> Enum.reduce(binding, fn {name, value}, binding ->
+            Keyword.put(binding, name, value)
+          end)
+
+        # Primitive heuristic to detect the used variables. This will not handle
+        # shadowing of variables in funs and will only work well enough for
+        # expressions, not for modules.
+        used_vars =
+          tokens
+          |> Enum.reduce(%{}, fn
+            {:var, _, name}, acc when not is_map_key(acc, name) ->
+              Map.put(acc, name, {erlang_to_elixir_var(name), nil})
+
+            _, acc ->
+              acc
+          end)
+          |> Map.values()
+          |> MapSet.new()
+
+        env =
+          update_in(
+            env.versioned_vars,
+            fn versioned_vars ->
+              versioned_vars
+              |> Map.filter(fn {name, _} -> MapSet.member?(used_vars, name) end)
+              |> Map.merge(
+                new_erl_binding
+                |> Map.keys()
+                |> Enum.map(&({&1, nil}))
+                |> Enum.with_index(Kernel.map_size(versioned_vars) + 1)
```

Comment:
```suggestion
                |> Enum.with_index(Kernel.map_size(versioned_vars) + 1, fn var, index -> {{var, nil}, index} end)
```

Reply:
This doesn't work in stable versions of Elixir, I'll revert.